app/gui/src/project-view/components/MarkdownEditor/codemirror/formatting/block.ts

```diff
@@ -1,285 +1,170 @@
+/** Apply the specified block type to the selection. */
+export function setBlockType(
+  state: EditorState,
+  type: DelimitedBlockType & SupportedBlockType,
+): TransactionSpec {
+  const selection = selectionRange(state.selection.main)
+  const md = new MarkdownEdit(state.doc, syntaxTree(state))
+  const delimiter = blockDelimiter[type]
+  const cursor = md.tree.cursor()
+  if (selection.empty && md.doc.lineAt(selection.from).length === 0)
+    md.replace(selection, delimiter)
+  visitLeafBlocks(selection, md.tree, (node, parentList) => {
+    if (node.name === type || (node.name === 'ListItem' && parentList === type)) return
+    const isQuote = node.name === 'Blockquote'
+    cursor.moveTo(node.from, 1)
+    for (;;) {
+      const oldDelimiter =
+        isBlockDelimiter(cursor.name) ? nodeRange(cursor) : Range.emptyAt(node.from)
+      md.replace(oldDelimiter, delimiter)
+      if (isQuote && cursor.nextSibling()) {
+        while (cursor.name !== 'QuoteMark') {
+          if (!cursor.nextSibling()) break
+        }
+        if (cursor.name === 'QuoteMark') continue
+      }
+      break
     }
+  })
+  const changes = state.changes(md.changes)
+  return {
+    changes,
+    selection: state.selection.main.map(changes),
   }
-  return changeSet
 }
 
-function findAtxHeaderMark(node: SyntaxNode): { level: number; from: number; to: number } | null {
-  const cursor = node.cursor()
-  do {
-    if (cursor.type.name.startsWith('ATXHeading')) {
-      const headerMark = cursor.node.getChild('HeaderMark')
-      if (!headerMark) return null
-      const level = Number(cursor.type.name.slice(-1))
-      return { level, from: headerMark.from, to: headerMark.to }
-    }
-  } while (cursor.parent())
-  return null
-}
-
-// =============
-// === Lists ===
-// =============
-
-/** Distinguish between unordered (bullet) and ordered (numbered) lists. */
-export type ListType = 'unordered' | 'ordered'
-
-/** Toggle list items of specified type at each of the selected lines. */
-export function toggleList(view: EditorView, type: ListType) {
-  const tree = ensoMarkdownParser.parse(view.state.doc.toString())
-  const startLine = view.state.doc.lineAt(view.state.selection.main.from)
-  const endLine = view.state.doc.lineAt(view.state.selection.main.to)
-  const changeSet = new MutableChangeSet(0)
-  const src = view.state.doc.toString()
-  let listIndex = 0
-  for (let i = startLine.number; i <= endLine.number; i++) {
-    const line = view.state.doc.line(i)
-    const context = new Context(tree, src, line)
-    const lineChanges = toggleListInner(context, listIndex, type)
-    changeSet.merge(lineChanges)
-    listIndex++
-  }
-  changeSet.dispatch(view, true)
+/**
+ * @returns Whether {@link insertCodeBlock} is supported for the current cursor location or
+ * selected range.
+ */
+export function canInsertCodeBlock(_state: EditorState): boolean {
+  // TODO: Disable button when the cursor is already inside an unformattable block.
```

Comment:
TODO left - when we would fix that?

Reply:
Implemented a solution that covers most cases.